weave/flow/casting.py

```diff
@@ -0,0 +1,54 @@
+from __future__ import annotations
+
+from typing import Annotated, Any, Union
+
+from pydantic import BeforeValidator
+
+import weave
+from weave.flow.dataset import Dataset
+from weave.scorers.base_scorer import Scorer, _validate_scorer_signature
+from weave.trace.op import Op, as_op, is_op
+from weave.trace.refs import ObjectRef, OpRef
+from weave.trace.vals import WeaveObject
+
+
+def cast_to_dataset(obj: Any) -> Dataset:
+    if isinstance(obj, Dataset):
+        return obj
+
+    if isinstance(obj, WeaveObject):
+        return Dataset.from_obj(obj)
+
+    if isinstance(obj, ObjectRef):
+        return obj.get()
+
+    if isinstance(obj, list):
+        return Dataset(rows=obj)
+
+    raise TypeError("Unable to cast to Dataset")
+
+
+def cast_to_scorer(obj: Any) -> Scorer | Op:
+    res: Scorer | Op
+    if isinstance(obj, Scorer):
+        res = obj
+    elif isinstance(obj, type):
+        raise TypeError(
+            f"Scorer {obj.__name__} must be an instance, not a class. Did you instantiate?"
+        )
+    elif callable(obj) and not is_op(obj):
+        res = weave.op(obj)
+    elif is_op(obj):
```

Comment:
If you swap this clause with the one before it seems like you could remove the `not is_op` check